WooCommerce/Classes/Authentication/Epilogue/StorePickerCoordinator.swift

```diff
@@ -0,0 +1,132 @@
+import Foundation
+import UIKit
+import Yosemite
+
+
+/// Simplifies and decouples the store picker from the caller
+///
+final class StorePickerCoordinator: Coordinator {
+
+    unowned var navigationController: UINavigationController
+
+    /// Determines how the store picker should initialized
+    ///
+    var selectedConfiguration: StorePickerConfiguration
+
+    /// Closure to be executed upon dismissal of the store picker
+    ///
+    var onDismiss: (() -> Void)?
+
+    /// Site Picker VC
+    ///
+    private lazy var storePicker: StorePickerViewController = {
+        let pickerVC = StorePickerViewController()
+        pickerVC.delegate = self
+        pickerVC.configuration = selectedConfiguration
+        return pickerVC
+    }()
+
+    init(_ navigationController: UINavigationController, config: StorePickerConfiguration) {
+        self.navigationController = navigationController
+        self.selectedConfiguration = config
+    }
+
+    func start() {
+        showStorePicker()
+    }
+}
+
+
+// MARK: - StorePickerViewControllerDelegate Conformance
+//
+extension StorePickerCoordinator: StorePickerViewControllerDelegate {
+
+    func willSelectStore(with storeID: Int, onCompletion: @escaping SelectStoreClosure) {
+        guard selectedConfiguration == .switchingStores else {
+            onCompletion()
+            return
+        }
+        guard storeID != StoresManager.shared.sessionManager.defaultStoreID else {
+            onCompletion()
+            return
+        }
+
+        logOutOfCurrentStore(onCompletion: onCompletion)
+    }
+
+    func didSelectStore(with storeID: Int) {
+        guard storeID != StoresManager.shared.sessionManager.defaultStoreID else {
+            return
+        }
+
+        finalizeStoreSelection(storeID)
+        presentStoreSwitchedNotice()
+        onDismiss?()
+    }
+}
+
+// MARK: - Private Helpers
+//
+private extension StorePickerCoordinator {
+
+    func showStorePicker() {
+        if selectedConfiguration == .standard {
+            navigationController.present(storePicker, animated: true)
+        } else {
+            navigationController.pushViewController(storePicker, animated: true)
+        }
+    }
+
+    func presentStoreSwitchedNotice() {
```

Comment:
I am starting to wonder if we should try to abstract out presenting Notices. We have code very similar to this in several places across the app (i.e. code that configures and presents a Notice), that is tightly coupled with the fact that we are using `Notice`. I am just thinking out loud here, but would it make sense to explore some kind of abstraction (let's call it `NoticeProducer`) modelled as a protocol, with an extension that implements presentation of Notices for view controllers (`extension NoticeProducer where self = UIViewController`) and provides "domain methods" to present user-facing alerts (for example `presentAlertWithTitle` `presentAlertWithTitle:retryClosure:`...)

My concern, even though it is a minor one, would be that if we wanted to use a different library, or we needed to change the way Notices are initialised, or we needed to make any change in general, we'd need to touch a lot of different classes / files

Reply:
This is a good point @ctarda — we do have Notice code scattered throughout the app! Would you mind writing up a new issue for this so we can track it?

Reply:
++ excellent observation